File: Javascript/1.Agosto/5.Jbelmud text-15Agosto.js
//todo Data una stringa, restituisci una versione confusa di quella stringa in cui ogni parola viene trasformata utilizzando i seguenti vincoli:

//! La prima e l'ultima lettera delle parole rimangono al loro posto.
//! Tutte le lettere tra la prima e l'ultima lettera vengono ordinate alfabeticamente.
//! Le stringhe di input non conterranno punteggiatura e saranno interamente in minuscolo.

//? Tempo di esecuzione:
//* Non concluso

//?Codice scritto senza utilizzo di ChatGpt

function jbelmu(text) {
  let firstChar = text[0]
  let lastChar = text.split("").splice(1).pop()
  let word = text.split(" ").splice(1)
  let wordRandom = word.sort((a,b) => b.localeCompare(a))

  return wordRandom;
}

console.log(jbelmu("hello world"))

//?Codice scritto da ChatGpt

function jumbleWords(s) {
  return s
    .split(" ")
    .map(word => {  //!Perchè pensare ad un forEach? Che ti ha fatto di male il map?
      if (word.length <= 2) return word;

      const middle = word
        .slice(1, -1)
        .split("")
        .sort()
        .join(""); //!Qui avevo capito come fare, ma mi sono perso sull'applciazione

      return word[0] + middle + word[word.length - 1];
    })
    .join(" ");
}

console.log(jumbleWords("coding is fun")); 

//* Hai provato a separare stringa e lettere

//* Stai usando split, sort, localeCompare (bene sperimentare)

//! Stai lavorando sulle parole, non sulle lettere interne

//! firstChar e lastChar non vengono mai usati

//! splice(1) rompe la struttura della parola

//! L’ordinamento richiesto è alfabetico delle lettere centrali, non delle parole

//! Non ricomponi la stringa finale corretta

//todo Considerazioni: 

//* Iniziamo con il dire che ho sbagliato approccio nei confronti dell'esercizio. 
//* Mi sono totalmente focalizzato sulla separazione della frase e non della parola, poichè come sempre non seguo bene le istruzioni.

//!N.B. MENO FRETTA E PIU' LOGICA. LEGGI BENE LA CONSEGNA

File: Python/1.Agosto/5.Jbelmud text-15Agosto.py
# ============================================================
# 🧠 ESERCIZIO
# ============================================================

# Data una stringa, restituisci una versione modificata in cui:
# - La prima e l'ultima lettera di ogni parola restano uguali
# - Le lettere centrali vengono ordinate alfabeticamente
# - Tutto è in minuscolo e senza punteggiatura

# Esempio:
# "jumbled" → "jbelmud"


# ============================================================
# 👨‍💻 CODICE SCRITTO DA ME
# ============================================================

def jbelmu(text):

    word_split = text.split()
    map_list = list(map(lambda w: w[0] + "".join(sorted(w[1:-1])) + w[-1] if len(w) > 1 else w, word_split))

    return " ".join(map_list)


print(jbelmu("i love jumbled text"))


# ============================================================
# ⚡ FEEDBACK VELOCE
# ============================================================

# Codice già funzionante e ben fatto 👍
# Hai usato correttamente:
# - split()
# - map()
# - lambda
# - sorted()

# Piccolo miglioramento sulla condizione


# ============================================================
# ❌ ERRORI SPIEGATI
# ============================================================

# ❌ len(w) > 1
# Problema:
# parole di 2 lettere ("hi") vengono comunque processate
# ma non serve modificarle

# ✔️ Meglio usare:
# len(w) > 2
# così modifichi solo parole con almeno 3 lettere


# ============================================================
# ✅ CODICE CORRETTO
# ============================================================

def jbelmu(text):

    # Divido la frase in parole
    word_split = text.split()

    # Trasformo ogni parola
    result = list(map(
        lambda w: w[0] + "".join(sorted(w[1:-1])) + w[-1] if len(w) > 2 else w,
        word_split
    ))

    # Ricompongo la frase
    return " ".join(result)


# ============================================================
# 📖 SPIEGAZIONE RIGA PER RIGA
# ============================================================

# def jbelmu(text):
# → funzione che prende una stringa

# word_split = text.split()
# → divide la frase in parole
# esempio: "i love text" → ["i", "love", "text"]

# map(...)
# → applica una funzione a ogni parola

# lambda w:
# → funzione veloce per lavorare su ogni parola

# w[0]
# → prima lettera

# w[1:-1]
# → lettere centrali
# esempio: "jumbled" → "umble"

# sorted(w[1:-1])
# → ordina le lettere centrali
# → ['b','e','l','m','u']

# "".join(...)
# → trasforma la lista in stringa
# → "belmu"

# w[-1]
# → ultima lettera

# risultato:
# "j" + "belmu" + "d" → "jbelmud"

# if len(w) > 2 else w
# → se parola corta ("i", "hi"), la lasci invariata

# " ".join(result)
# → ricompone la frase finale


# ============================================================
# 🧪 TEST
# ============================================================

print(jbelmu("i love jumbled text"))

# Output:
# "i love jbelmud tettx"